### DISCUSSION TITLE
Response compression stopped working sometime between 5.3.104 and 5.4.5

### DISCUSSION THREAD

**Discussion by krivulcik** [ravendb/ravendb#15621] [Q&A] [answered]
We upgraded our clusters from 5.3.104 to 5.4.5 in December. We have noticed a significant increase in traffic after the upgrade.

The response compression seems to have stopped working sometime between those two versions, seemingly at different times for HTTP and HTTPS traffic (with `Http.AllowResponseCompressionOverHttps` set to true).

I've tested versions 5.3.104 (our previous version), 5.3.108 (the latest 5.3 version), 5.4.5 (our current version), and 5.4.1 (the earliest 5.4 version).

In our production environment, lazy requests seem to be affected, so that's what I have focused on.

The results of the testing are the following. I tested with client version 5.4.5:

```
Version   compression used with HTTPS   compression used with HTTP
5.3.104   yes                           yes
5.3.108   yes                           NO
5.4.1     yes                           NO
5.4.5     NO                            NO
```

The requests loading two 5 MB documents look like this in Fiddler:

Compressed response (also, the stats show the response size of around 44 kB):
![image](https://user-images.githubusercontent.com/7997055/212107324-164cb606-58a2-49b3-8244-60c299fb19a7.png)

```
Request Count:   1
Bytes Sent:      416		(headers:251; body:165)
Bytes Received:  44 597		(headers:232; body:44 365)

ACTUAL PERFORMANCE
--------------
ClientConnected:	16:05:19.056
ClientBeginRequest:	16:05:19.868
GotRequestHeaders:	16:05:19.868
ClientDoneRequest:	16:05:19.868
Determine Gateway:	0ms
DNS Lookup: 		0ms
TCP/IP Connect:	2441ms
HTTPS Handshake:	0ms
ServerConnected:	16:05:22.310
FiddlerBeginRequest:	16:05:22.310
ServerGotRequest:	16:05:22.310
ServerBeginResponse:	16:05:22.313
GotResponseHeaders:	16:05:22.313
ServerDoneResponse:	16:05:22.359
ClientBeginResponse:	16:05:22.359
ClientDoneResponse:	16:05:22.359

	Overall Elapsed:	0:00:02.491

RESPONSE BYTES (by Content-Type)
--------------
application/json: 44 365
       ~headers~: 232
```

Uncompressed response (stats show the response size of 10 MB):
![image](https://user-images.githubusercontent.com/7997055/212107810-c00977b6-60fc-4a8c-92ef-548d4fb5ef89.png)

```
Request Count:   1
Bytes Sent:      416		(headers:251; body:165)
Bytes Received:  10 000 924		(headers:206; body:10 000 718)

ACTUAL PERFORMANCE
--------------
ClientConnected:	16:05:19.056
ClientBeginRequest:	16:05:32.217
GotRequestHeaders:	16:05:32.217
ClientDoneRequest:	16:05:32.217
Determine Gateway:	0ms
DNS Lookup: 		0ms
TCP/IP Connect:	2031ms
HTTPS Handshake:	0ms
ServerConnected:	16:05:34.248
FiddlerBeginRequest:	16:05:34.248
ServerGotRequest:	16:05:34.248
ServerBeginResponse:	16:05:34.268
GotResponseHeaders:	16:05:34.268
ServerDoneResponse:	16:05:34.309
ClientBeginResponse:	16:05:34.316
ClientDoneResponse:	16:05:34.320

	Overall Elapsed:	0:00:02.103

RESPONSE BYTES (by Content-Type)
--------------
application/json: 10 000 718
       ~headers~: 206
```

Versions 5.3.108 and 5.4.1 behave differently depending on whether HTTPS is used. For example, 5.4.1:

Plain HTTP:
![image](https://user-images.githubusercontent.com/7997055/212108909-578e6db6-c3d2-4d60-8a72-00167a0d9d8d.png)

```
Request Count:   1
Bytes Sent:      416		(headers:251; body:165)
Bytes Received:  44 593		(headers:230; body:44 363)

ACTUAL PERFORMANCE
--------------
ClientConnected:	16:05:19.056
ClientBeginRequest:	16:05:30.044
GotRequestHeaders:	16:05:30.044
ClientDoneRequest:	16:05:30.044
Determine Gateway:	0ms
DNS Lookup: 		0ms
TCP/IP Connect:	2025ms
HTTPS Handshake:	0ms
ServerConnected:	16:05:32.070
FiddlerBeginRequest:	16:05:32.070
ServerGotRequest:	16:05:32.070
ServerBeginResponse:	16:05:32.072
GotResponseHeaders:	16:05:32.072
ServerDoneResponse:	16:05:32.129
ClientBeginResponse:	16:05:32.129
ClientDoneResponse:	16:05:32.129

	Overall Elapsed:	0:00:02.084

RESPONSE BYTES (by Content-Type)
--------------
application/json: 44 363
       ~headers~: 230
```

Encrypted HTTPS:
![image](https://user-images.githubusercontent.com/7997055/212109187-1380a91b-f674-4479-b44d-2789a1c57eb7.png)

```
Request Count:   1
Bytes Sent:      445		(headers:280; body:165)
Bytes Received:  10 000 925		(headers:183; body:10 000 742)

ACTUAL PERFORMANCE
--------------
ClientConnected:	16:10:45.594
ClientBeginRequest:	16:10:45.767
GotRequestHeaders:	16:10:45.767
ClientDoneRequest:	16:10:45.767
Determine Gateway:	0ms
DNS Lookup: 		0ms
TCP/IP Connect:	0ms
HTTPS Handshake:	0ms
ServerConnected:	16:10:45.631
FiddlerBeginRequest:	16:10:45.767
ServerGotRequest:	16:10:45.767
ServerBeginResponse:	16:10:45.962
GotResponseHeaders:	16:10:45.962
ServerDoneResponse:	16:10:51.237
ClientBeginResponse:	16:10:51.239
ClientDoneResponse:	16:10:51.419

	Overall Elapsed:	0:00:05.651

RESPONSE BYTES (by Content-Type)
--------------
application/json: 10 000 742
       ~headers~: 183

```

Was there a regression, or is this an intended/expected change? Should we do something differently to have the responses compressed?

Testing code:

``` c#
using Raven.Client.Documents;
using System;
using System.Security.Cryptography.X509Certificates;

namespace RavenDB5Tests.Tryouts
{
    public static class Compression
    {
        public static void CompressionTest(string url, X509Certificate2 certificate)
        {
            using (var store = new DocumentStore
            {
                Urls = new[] { url },
                Certificate = certificate,
                Database = "compression-test",
                Conventions = new Raven.Client.Documents.Conventions.DocumentConventions
                {
                    DisableTopologyUpdates = true,
                }
            }.Initialize())
            {
                // uncomment for the first run
                //using (var session = store.OpenSession())
                //{
                //    var doc = new Doc { Id = "doc-1", StrVal = new string('a', 5000000) };
                //    session.Store(doc);
                //    var doc2 = new Doc { Id = "doc-2", StrVal = new string('b', 5000000) };
                //    session.Store(doc2);
                //    session.SaveChanges();
                //}

                using (var session = store.OpenSession())
                {
                    Console.WriteLine("Loading.");
                    var doc1Lazy = session.Advanced.Lazily.Load<Doc>("doc-1");
                    var doc2Lazy = session.Advanced.Lazily.Load<Doc>("doc-2");
                    var doc1 = doc1Lazy.Value;
                    var doc2 = doc2Lazy.Value;
                    Console.WriteLine("Loaded.");
                }
            }
        }

        public class Doc
        {
            public string Id { get; set; }
            public string StrVal { get; set; }
        }
    }
}
```

I call it like this:
```
            X509Certificate2 clientCertificate = new X509Certificate2("my-certificate.pfx");

            var httpServers = new[]
            {
                "http://localhost:44312/", // 5.3.104
                "http://localhost:44314/", // 5.3.108
                "http://localhost:44315/", // 5.4.1
                "http://localhost:44313/", // 5.4.5
            };

            var httpsServers = new[]
            {
                "https://a.testing-domain-for-53104.development.run:44312/", // 5.3.104
                "https://a.testing-domain-for-53108.development.run:44314/", // 5.3.108
                "https://a.testing-domain-for-541.development.run:44315/", // 5.4.1
                "https://a.testing-domain-for-545.development.run:44313/", // 5.4.5
            };

            //foreach (var url in httpServers)
            //{
            //    Compression.CompressionTest(url, null);
            //}

            foreach (var url in httpsServers)
            {
                Compression.CompressionTest(url, clientCertificate);
            }
```

**Comment by maciejaszyk** [comment]
Hey @krivulcik,
I tried this on: `5.3.108` and `5.4.5` and HTTP seems to be working just fine.
![image](https://user-images.githubusercontent.com/86351904/212319990-e7782eb4-dea4-41c4-85f0-b62b2a9336a3.png)


I have a problem with Fiddler when looking into secured Raven's instance. I see only tunneling reqs, have you changed anything in the configuration to be able to get HTTPS `multi_get` requests?

**Reply by krivulcik** [reply]
You need to configure Fiddler to send the client certificate. I configure it like this:

```
		if (oSession.uriContains("whatever.development.run"))
		{
			oSession["https-Client-Certificate"] = "c:\\Users\\andrej\\Downloads\\RavenDB\\andrej.cer";
		}
```

You need to have the certificate and the key in your system store. The .cer file only serves as a pointer to that certificate.

I put this at the beginning of `static function OnBeforeRequest(oSession: Session)` in the Fiddler rules (Rules -> Customize Rules (Ctrl-R)).

See also: https://docs.telerik.com/fiddler/configure-fiddler/tasks/respondwithclientcert

I'll re-check 5.4.5 with plain HTTP to check if there might be something wrong with my setup. But in our production environment, the requests are not compressed. The environment is Windows Server VMs in Azure.

**Reply by maciejaszyk** [reply]
BTW:
Do you use antivirus on your Raven server or app server? Sometimes it decodes the message.

**Reply by krivulcik** [reply]
These tests were done on my laptop, with only Windows Defender. I don't think that it should intercept traffic (and is unable to do so in the case of HTTPS traffic), and even if it would, there shouldn't be a difference between the RavenDB versions.

**Reply by krivulcik** [reply]
I just re-checked version 5.4.5, just to be sure.

This is the request/response for plain HTTP with RavenDB 5.4.5 on my laptop:

```
Request Count:   1
Bytes Sent:      416		(headers:251; body:165)
Bytes Received:  10 000 924		(headers:206; body:10 000 718)

ACTUAL PERFORMANCE
--------------
ClientConnected:	16:55:47.377
ClientBeginRequest:	16:55:47.532
GotRequestHeaders:	16:55:47.532
ClientDoneRequest:	16:55:47.532
Determine Gateway:	0ms
DNS Lookup: 		0ms
TCP/IP Connect:	2044ms
HTTPS Handshake:	0ms
ServerConnected:	16:55:49.578
FiddlerBeginRequest:	16:55:49.578
ServerGotRequest:	16:55:49.579
ServerBeginResponse:	16:55:49.656
GotResponseHeaders:	16:55:49.656
ServerDoneResponse:	16:55:49.690
ClientBeginResponse:	16:55:49.696
ClientDoneResponse:	16:55:49.700

	Overall Elapsed:	0:00:02.167

RESPONSE BYTES (by Content-Type)
--------------
application/json: 10 000 718
       ~headers~: 206
```

![image](https://user-images.githubusercontent.com/7997055/212363267-5e3ebb81-6599-49a8-9c17-f52a774bf4d6.png)

The response is uncompressed. Server/client versions are the same. The other versions, including 5.3.108 (from your screenshot) work correctly - the response is compressed.

**Comment by maciejaszyk** [comment, marked as answer]
Hi @krivulcik .
We found the issue! :) I've prepared a fix for this:
https://github.com/ravendb/ravendb/pull/15656

Thank you for reporting this. :)

BTW: Do you need a custom build with this fix?

**Reply by krivulcik** [reply]
Thanks, @maciejaszyk. If there's a stable release scheduled in a week or so, we can wait for that. But if we'd need to wait longer, a build based on a stable build with this fix incorporated would be appreciated.

**Reply by maciejaszyk** [reply]
We made a custom build for you. This is 5.4.100 + fix

windows: https://daily-builds.s3.us-east-1.amazonaws.com/RavenDB-5.4.100-custom-54029-windows-x64.zip

linux: https://daily-builds.s3.us-east-1.amazonaws.com/RavenDB-5.4.100-custom-54029-linux-x64.tar.bz2
